nixos/modules/services/web-apps/hedgedoc.nix

```diff
@@ -1018,58 +271,99 @@ in
         `HedgeDoc` is running.
       '';
     };
-
-    package = mkOption {
-      type = types.package;
-      default = pkgs.hedgedoc;
-      defaultText = literalExpression "pkgs.hedgedoc";
-      description = lib.mdDoc ''
-        Package that provides HedgeDoc.
-      '';
-    };
-
   };
 
   config = mkIf cfg.enable {
-    assertions = [
-      { assertion = cfg.settings.db == {} -> (
-          cfg.settings.dbURL != "" && cfg.settings.dbURL != null
-        );
-        message = "Database configuration for HedgeDoc missing."; }
-    ];
-    users.groups.${name} = {};
+    users.groups.${name} = { };
     users.users.${name} = {
       description = "HedgeDoc service user";
       group = name;
-      extraGroups = cfg.groups;
-      home = cfg.workDir;
-      createHome = true;
       isSystemUser = true;
     };
 
-    systemd.services.hedgedoc = {
-      description = "HedgeDoc Service";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "networking.target" ];
-      preStart = ''
-        ${pkgs.envsubst}/bin/envsubst \
-          -o ${cfg.workDir}/config.json \
-          -i ${prettyJSON cfg.settings}
-        mkdir -p ${cfg.settings.uploadsPath}
-      '';
-      serviceConfig = {
-        WorkingDirectory = cfg.workDir;
-        StateDirectory = [ cfg.workDir cfg.settings.uploadsPath ];
-        ExecStart = "${cfg.package}/bin/hedgedoc";
-        EnvironmentFile = mkIf (cfg.environmentFile != null) [ cfg.environmentFile ];
-        Environment = [
-          "CMD_CONFIG_FILE=${cfg.workDir}/config.json"
-          "NODE_ENV=production"
-        ];
-        Restart = "always";
-        User = name;
-        PrivateTmp = true;
-      };
+    services.hedgedoc.settings = {
+      defaultNotePath = mkDefault "${cfg.package}/public/default.md";
+      docsPath = mkDefault "${cfg.package}/public/docs";
+      viewPath = mkDefault "${cfg.package}/public/views";
     };
+
+    systemd.services.hedgedoc =
+      let
+        # Some settings that we set as null will break hedgedoc,
+        # as it expects them either to be undefined or set.
+        cleanConfig = pkgs.runCommandLocal "hedgedoc-config.json"
+          {
+            nativeBuildInputs = [ pkgs.jq ];
+          } ''
+          jq '{production:del(.[]|nulls)|del(.[][]?|nulls)}' \
+            < ${settingsFormat.generate "hedgedoc-ugly.json" cfg.settings} \
+            > $out
+        '';
```

Comment:
We can set defaults for free form settings in config which is a lot cleaner

Reply:
The problematic options that gave rise to this in the first place were the SSL files, which AFAIK don't have a sensible default. But I see you've addressed that later on, so I'll leave this for now.

https://github.com/NixOS/nixpkgs/pull/244941#discussion_r1272739734